forc-plugins/forc-client/src/op/call.rs

```diff
@@ -0,0 +1,2221 @@
+use crate::{
+    cmd,
+    constants::DEFAULT_PRIVATE_KEY,
+    util::{
+        node_url::get_node_url,
+        tx::{prompt_forc_wallet_password, select_local_wallet_account},
+    },
+};
+use anyhow::{anyhow, bail, Result};
+use either::Either;
+use fuel_abi_types::abi::unified_program::UnifiedProgramABI;
+use fuels::{
+    crypto::SecretKey,
+    programs::calls::{receipt_parser::ReceiptParser, traits::TransactionTuner, ContractCall},
+};
+use fuels_accounts::{provider::Provider, wallet::WalletUnlocked};
+use fuels_core::{
+    codec::{encode_fn_selector, ABIDecoder, ABIEncoder, DecoderConfig, EncoderConfig},
+    types::{
+        bech32::Bech32ContractId,
+        param_types::ParamType,
+        transaction::TxPolicies,
+        transaction_builders::{BuildableTransaction, ScriptBuildStrategy, VariableOutputPolicy},
+        ContractId, EnumSelector, StaticStringToken, Token, U256,
+    },
+};
+use std::{collections::HashMap, fs::File, io::Write, str::FromStr};
+use sway_ast::{FnArgs, Item, ItemKind, Ty};
+use swayfmt::parse::with_handler;
+
+/// A command for calling a contract function.
+pub async fn call(cmd: cmd::Call) -> anyhow::Result<String> {
+    let cmd::Call {
+        contract_id,
+        abi,
+        function,
+        args,
+        node,
+        caller,
+        mode,
+        gas,
+        external_contracts,
+    } = cmd;
+    let node_url = get_node_url(&node, &None)?;
+    let provider: Provider = Provider::connect(node_url.clone()).await?;
+
+    let wallet = get_wallet(caller, provider).await?;
+
+    if let Some(abi) = abi {
+        // If ABI is provided, ensure function signature is just the selector
+        let cmd::call::FuncType::Selector(selector) = function else {
+            bail!("Function must be a selector");
+        };
+
+        let (file_path, is_temp_file) = match abi {
+            Either::Left(path) => (path, false),
+            Either::Right(url) => {
+                // Download the file to tempfile; move blocking operations to a new thread
+                let path = std::env::temp_dir().join("temp_abi.json");
+                let mut file = File::create(&path)?;
+                let response = reqwest::get(url).await?.bytes().await?;
+                file.write_all(&response)?;
+                (path, true)
+            }
+        };
+
+        let abi_str = std::fs::read_to_string(file_path.clone()).expect("Failed to read ABI file");
+        if is_temp_file {
+            std::fs::remove_file(file_path).expect("Failed to remove ABI file");
+        }
+        let parsed_abi = UnifiedProgramABI::from_json_abi(&abi_str)?;
```

Comment:
should we add validation to ensure ABI files are well-formed and compatible before progressing?

```rust
// Validate minimum required fields
if parsed_abi.functions.is_empty() {
    bail!("Invalid ABI: No functions found in ABI file");
}

...
```

Reply:
I was thinking we may want to support calling a contract without any functions - where you may want to only specify an amount.
In the case of solidity, this calls the default `fallback` or `receive` function - allowing the contract to receive funds.

I was unsure whether sway has an equivalent default function - or an approach where the contract can receive funds without calling an explicit function.

Reply:
This is indeed possible, rather than using a `CALL` instruction, you can use a [`TR` or transfer instruction](https://docs.fuel.network/docs/specs/fuel-vm/instruction-set/#tr-transfer-coins-to-contract) to send funds into a contract.

https://docs.fuel.network/docs/nightly/fuels-ts/wallets/wallet-transferring/#transferring-assets-to-contracts

The sway also has a notion of default functions which are used by proxy contracts to pass-through calls and args to an implementation contract, but its not necessary to use this default function to receive funds.